mara_db/bigquery.py

```diff
@@ -4,30 +4,171 @@
     except Exception as e:
         connection.close()
         raise e
-    finally:
-        cursor.close()
-        connection.close()
+
+
+def create_bigquery_table_from_postgresql_query(
+        postgresql_query: str, postgresql_db_alias: str,
+        bigquery_db_alias: str, bigquery_dataset_id: str, bigquery_table_name: str):
+    """
+    Creates a table in for bigquery from a Postgresql SELECT query. Will print the query
+
+    Usefull for copying PostgreSQL tables to BigQuery (create table first and then copy)
+
+    Example:
+        >>> create_bigquery_table_from_postgresql_query(
+        >>>              postgresql_db_alias='dwh',
+        >>>              postgresql_query='SELECT 1::SMALLINT AS a, now() as b',
+        >>>              bigquery_db_alias='reporting',
+        >>>              bigquery_dataset_id='foo',
+        >>>              bigquery_table_name='bar')
+        CREATE OR REPLACE TABLE `foo`.`bar` (
+            `a` INT64,
+            `b` TIMESTAMP
+        )
+
+    Args:
+        postgresql_query: The query to execute in PostgreSQL, must not end with a semicolon
+        postgresql_db_alias: The postgresql database to execute the query in
+        bigquery_db_alias: The mara db alias of the bigquery connection
+        bigquery_dataset_id: The id of the bigquery dataset in which the table is to be created
+        bigquery_table_name: The name of the to be created table
+    """
+    from mara_db.postgresql import postgres_cursor_context
+    with mara_db.postgresql.postgres_cursor_context(postgresql_db_alias) as cursor:
+        cursor.execute('SELECT oid, typname FROM pg_type;')
+        pg_types = {}
+        for oid, type_name in cursor.fetchall():
+            pg_types[oid] = type_name
+
+        # https://cloud.google.com/bigquery/docs/reference/standard-sql/federated_query_functions#postgressql_mapping
+        pg_to_bigquery_type_mapping = {
+            'bool': 'BOOL',
+            'bytea': 'BYTES',
+            'date': 'DATE',
+            'int2': 'INT64',
+            'int4': 'INT64',
+            'int8': 'INT64',
+            'json': 'STRING',
+            'jsonb': 'STRING',
+            'numeric': 'NUMERIC',
+            'float4': 'FLOAT64',
+            'float8': 'FLOAT64',
+            'varchar': 'STRING',
+            'text': 'STRING',
+            'time': 'TIME',
+            'timestamp': 'DATETIME',
+            'timestamptz': 'TIMESTAMP',
+        }
+
+        cursor.execute(postgresql_query + ' LIMIT 0')
+
+        column_specs = []
+        for column in cursor.description:
+            pg_type = pg_types[column.type_code]
+            assert pg_type in pg_to_bigquery_type_mapping, f"Unmapped type '{pg_type}'"
+            column_specs.append(f'`{column.name}` {pg_to_bigquery_type_mapping[pg_type]}')
+
+        query = f"""
+CREATE OR REPLACE TABLE `{bigquery_dataset_id}`.`{bigquery_table_name}` (
+    """ + ',\n    '.join(column_specs) + "\n)"
+
+        print(query)
```

Comment:
Is this print intended?

Reply:
Yes, the query is created at run-time and otherwise you don't see what's wrong when there's and exception